### ISSUE TITLE
The performance for VOC07 Supervised Learning

### ISSUE THREAD

**Issue by machengcheng2016** [soo89/ISD-SSD#5]
Greetings!
I notice that the batch size is always 32 in `train_ssd.py`, `train_csd.py` and `train_isd.py`. For supervised training, the 32 images in one batch are all labeled. For semi-supervised learning, some images in one batch are labeled and the rest in the batch are unlabeled. 

When I manually set the unsupervised learning loss as `0` in `train_csd.py` and `train_isd.py`, the unlabeled images in one batch will make no difference to the training process, so we can regard the training process as supervised. Then the batch size is actually the number of labeled images in one batch, which is surely less than 32. After 120k iterations, I find that the mAP is definitely larger than `70.2`. I've repeated this experiment five times, and the mAPs are `71.4`, `71.5`, `71.7`, `71.7`, `72.3`. 
As a result, we can find that both CSD and ISD seems less effective, especially for CSD (even no improvement on mAP). How do you think? 
By the way, why don't you adopt the data loader type in your CSD-RFCN? That `1:3` data loader seems a fair experimental setting. And not surprisingly, the CSD method only improve 0.8 mAP on RFCN as shown in your CSD paper.

**Comment by soo89** [comment]
We followed the setting of Semi-supervised learning in classification.

I haven't experimented with the unsupervised loss learning as zero as you did, so you might be right.

However, I followed the conventional method of semi-supervised learning, which compared to supervised learning that uses full labeled data in mini-batch.

I think Randomizing the number of samples in mini-batch or other things can give a good effect.

**Comment by soo89** [comment]
R-FCN requires much memory, so we could only train with a batch size of 4. If we use an SSD data loader, there would be many cases without labels in mini-batch. So, we use a different data loader.

**Comment by soo89** [comment]
If possible, can you check out the SSD512 model as well?
I'm also curious.

**Comment by soo89** [comment]
We followed the setting of conventional semi-supervised learning.

They trained with full-labeled data in a mini-batch when they train a model under a supervised training setting.

And, when we train a model under a supervised learning setting, our model (70.2) shows a better score than the SSD paper (68.0).

For the data loader, I also followed the SSD's semi-supervised learning settings. However, since the batch size of R-FCN is 4, the mini-batch may not have labeled samples. So we created a new data loader.
In other words, the data loader of SSD is the same as a semi-supervised learning setting of classification, and the data loader of R-FCN is the method we applied.

**Comment by soo89** [comment]
We also tested that setting that you mentioned. 

And we got 70.8 mAP in our environment. 

As you mentioned, it shows better than 70.2, but it is lower than what you mentioned. 

Please check it again.

**Comment by machengcheng2016** [comment]
As I validated five times, the mAP of VOC07 supervised learning is ranging from `71.4` to `72.3` when setting the CSD/ISD loss term to zero (in other words, keep the same batch size with your CSD/ISD).
Besides, I have tried CSD twice, and the mAPs are `72.10`, `72.08`. Compare to the correct baseline performance (`72.3`), the proposed CSD is absolutely useless and make no difference to semi-supervised learning.

By mentioning RFCN, I just suggest you changing the type of data loader you currently use to that type in RFCN. In fact I have done this. Here is the experimental results.
Epoch: 200
VOC07 supervised (32 labeled): `69.75`
CSD (32 labeled + 96 unlabeled): `70.28`
CSD only gets ~0.5 mAP improvements.

Thank you.

**Comment by soo89** [comment]
I experimented last night and checked it.

I wonder if you experimented in the correct setting.

You should respect other researchers.

**Comment by soo89** [comment]
I am very offended by your discussion attitude. 

How can you be so sure your result is absolutely right and mine is not.

After taking your comment, I  even ran an extra experiment to show you the results.

Did you even check your code again as I advised?

No matter how many times you run the test, you will never get an accurate result with an incorrect setting.

Why did you change the hyperparameter of R-FCN as you like?

It is also important to set the experiment according to the conventional method, and I followed the classification semi-supervised learning setting. 

I think there are two possible cases when you get that score.

1. For supervised learning, you need to change the code of 'train_csd.py' as follows.
(L319-325)


        if(supervised_flag ==1):	
                loss = loss_l + loss_c 
        else:
                if(len(sup_image_index)==0):
                        continue
                else:
                        loss = loss_l + loss_c 


For the initial stable training, I train the model with only labeled data for 1 epoch and then train with a semi-supervised learning setting.

If you only changed the loss in supervised_flag, you trained a model under the semi-supervised setting, it makes sense to have a score of 72.3.

Second, the file may have been overwritten.

Results of Object detection don’t have much variation. 

I doubt that your models are overwritten with every new training.

**Comment by machengcheng2016** [comment]
Greetings.
1. As for supervised learning, I just change the code of `train_csd.py` as follows.
(L291)
`if(mask_val.sum()>0):`
->
`if False`

Then I tried five times and get mAPs ranging from `71.4` to `72.3`.

2. As for the CSD. Not just me. There is still [another paper](https://arxiv.org/abs/2103.16368) reporting the real performance of CSD: `72.3`

3. You are saying you've done everything right, but I notice that you modified the layers/modules/isd_loss.py 25 days ago. You did change L35 from `right_conf_class = conf_shuffle[:, :, 1:].clone()` to `right_conf_class = conf_temp[:, :, 1:].clone()`, which means you had been using incorrect mask for the right image in ISD loss before the modification. I really doubt the reliability of your CSD/ISD experimental results.

**Comment by soo89** [comment]
1. You have experimented without BE algorithm, not a supervised learning setting.

You experimented incorrectly.

2. In other paper, they mentioned our semi-supervised learning setting(72.3) in Table 1, and it is correct. 

3. I found our mistake, and we have fixed the error and experimented again. Therefore, our scores of the paper (v1 and v2 in arxiv) are different.

**Comment by machengcheng2016** [comment]
1. Don't be mad. Please take a second look at the `train_csd.py` and see what is going on after changing L291.
2. Glad you admit your CSD result is actually `72.3`, not `72.7` as reported in your CSD paper. So the effect of CSD loss is quite limited.
3. Okay. Good for you!

**Comment by soo89** [comment]
2. where is the 72.7  ?? 

I have experimented several times and have even gotten 72.7.

Nevertheless, I reported in the paper as 72.3.

Please check the paper in table1.

https://papers.nips.cc/paper/2019/hash/d0f4dae80c3d0277922f8371d5827292-Abstract.html

**Comment by machengcheng2016** [comment]
Alright. What about my question 1?

**Comment by soo89** [comment]
With your setting, we got 72.0.

you can check in 5 rows in table 1. (cls+loc , not BE)

**Comment by machengcheng2016** [comment]
You mean `consistency_conf_loss` and `consistency_loc_loss` are not zero? Would you take a second look at the `train_csd.py`, please?

**Comment by soo89** [comment]
I confused with my code. 

I will train with your setting. 

However, when I erase the variable in the loss, I got 70.8.

I think it should be the same, but I will check.

**Comment by machengcheng2016** [comment]
Cool. As long as you calm down and be willing to check your code :)

**Comment by soo89** [comment]
I apologize for losing my temper.

I experimented and convinced my results. Nevertheless, I thought you ignored my result.

I experimented three times on your experimental settings and got 71.7, 71.3, and 70.4 mAP.

I am very surprised by the 71.7 mAP result.

And I thought about the baseline of supervised learning.

I still insist that the baseline of supervised learning is 70.2.

In conventional supervised learning, we do not experiment by changing the number of samples in the mini-batch. 

What we train in the same way as conventional supervised learning should be the baseline.

However, it can be said that the additional effect of CSD training (changing the number of samples in supervised loss) can have a significant impact on performance.
(To my best knowledge, the different number of samples in a mini-batch changes the weight for each object sample to be more varied. And it makes the large variance as I have. )

I am sorry again, and thanks to you, I have learned that the above effects can be significant.
